librz/main/rz-diff.c

```diff
@@ -1,1305 +1,1454 @@
-				ctx = -1;
-				continue;
-			}
-		} else {
-			ctx = DUMP_CONTEXT;
-		}
-		rz_cons_printf(eq ? Color_GREEN : Color_RED);
-		rz_cons_printf("0x%08x%c ", i, eq ? ' ' : '!');
-		rz_cons_printf(Color_RESET);
-		for (j = 0; j < w; j++) {
-			bool eq2 = a[i + j] == b[i + j];
-			if (!eq) {
-				rz_cons_printf(eq2 ? Color_GREEN : Color_RED);
-			}
-			rz_cons_printf("%02x", a[i + j]);
-			if (!eq) {
-				rz_cons_printf(Color_RESET);
-			}
-		}
-		for (j = 0; j < pad; j++) {
-			rz_cons_printf("  ");
-		}
-		rz_cons_printf(" ");
-		for (j = 0; j < w; j++) {
-			bool eq2 = a[i + j] == b[i + j];
-			if (!eq) {
-				rz_cons_printf(eq2 ? Color_GREEN : Color_RED);
-			}
-			rz_cons_printf("%c", IS_PRINTABLE(a[i + j]) ? a[i + j] : '.');
-			if (!eq) {
-				rz_cons_printf(Color_RESET);
-			}
-		}
-		for (j = 0; j < pad; j++) {
-			rz_cons_printf(" ");
-		}
-		rz_cons_printf("   ");
-		for (j = 0; j < w; j++) {
-			bool eq2 = a[i + j] == b[i + j];
-			if (!eq) {
-				rz_cons_printf(eq2 ? Color_GREEN : Color_RED);
-			}
-			rz_cons_printf("%02x", b[i + j]);
-			if (!eq) {
-				rz_cons_printf(Color_RESET);
-			}
-		}
-		for (j = 0; j < pad; j++) {
-			rz_cons_printf("  ");
-		}
-		rz_cons_printf(" ");
-		for (j = 0; j < w; j++) {
-			bool eq2 = a[i + j] == b[i + j];
-			if (!eq) {
-				rz_cons_printf(eq2 ? Color_GREEN : Color_RED);
-			}
-			rz_cons_printf("%c", IS_PRINTABLE(b[i + j]) ? b[i + j] : '.');
-			if (!eq) {
-				rz_cons_printf(Color_RESET);
-			}
-		}
-		rz_cons_printf("\n");
-		rz_cons_flush();
+
+	if (!rz_core_bin_load(core, NULL, UT64_MAX)) {
+		rz_diff_error("cannot load file '%s'\n", filename);
+		goto rz_diff_load_file_with_core_fail;
 	}
-	rz_cons_break_end();
-	rz_cons_printf("\n" Color_RESET);
-	rz_cons_flush();
-	if (as != bs) {
-		rz_cons_printf("...\n");
+
+	if (!rz_core_bin_update_arch_bits(core)) {
+		rz_diff_error("cannot set architecture with bits\n");
+		goto rz_diff_load_file_with_core_fail;
 	}
+
+	bfile = core_get_file(cfile);
+	if (!bfile) {
+		rz_diff_error("cannot get architecture with bits\n");
+		goto rz_diff_load_file_with_core_fail;
+	}
+
+	RzList *sections = rz_bin_maps_of_file_sections(bfile);
+	if (rz_list_empty(sections)) {
+		rz_config_set_i(core->config, "io.va", false);
+	}
+	rz_list_free(sections);
+
+	if (architecture) {
+		rz_config_set(core->config, "asm.arch", architecture);
+	}
+
+	if (arch_bits) {
+		rz_config_set_i(core->config, "asm.bits", arch_bits);
+	}
+
+	if (!rz_core_analysis_everything(core, false, NULL)) {
+		rz_diff_error("cannot set analyze binary '%s'\n", filename);
+		goto rz_diff_load_file_with_core_fail;
+	}
+
+	return cfile;
+
+rz_diff_load_file_with_core_fail:
+	rz_core_free(core);
+	return NULL;
 }
 
-static void dump_cols_hexii(ut8 *a, int as, ut8 *b, int bs, int w) {
-	bool spacy = false;
-	ut32 sz = RZ_MIN(as, bs);
-	ut32 i, j;
-	int ctx = DUMP_CONTEXT;
-	int pad = 0;
-	if (!a || !b || as < 0 || bs < 0) {
-		return;
+static bool rz_diff_file_open(DiffFile *dfile, const char *filename) {
```

Comment:
probably worth to add `rz_return_val_if_fail()` here

Reply:
that can't happen